src/qml/qml_owned_ptr.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#ifdef MIXXX_DEBUG_ASSERTIONS_ENABLED
+#include <QObject>
+#include <QQmlEngine>
+#endif
+#include <QPointer>
+#include <memory>
+
+#include "util/assert.h"
+
+// Use this wrapper class to clearly represent a raw pointer that is owned by a
+// QML Engine. Objects which derive from QObject, have their lifetime governed
+// by the QML (or JavaScript) Engine, and thus such pointers do not require a
+// manual delete to free the heap memory when they go out of scope, as they will
+// be handled by the engine garbage collector.
+template<typename T>
+class qml_owned_ptr final {
+  public:
+    qml_owned_ptr() noexcept
+            : m_ptr{nullptr} {
+    }
+
+    qml_owned_ptr(std::nullptr_t) noexcept
+            : m_ptr{nullptr} {
+    }
+
+    explicit qml_owned_ptr(T* t) noexcept
+            : m_ptr{t} {
+        DEBUG_ASSERT(m_ptr);
+        QQmlEngine::setObjectOwnership(m_ptr, QQmlEngine::JavaScriptOwnership);
+    }
+
+    // explicitly generate trivial destructor (since decltype(m_ptr) is not a class type)
+    ~qml_owned_ptr() noexcept = default;
+
+    // Rule of 5
+    qml_owned_ptr(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&) = delete;
+    qml_owned_ptr(const qml_owned_ptr<T>&& other) = delete;
+    qml_owned_ptr& operator=(const qml_owned_ptr<T>&& other) = delete;
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is convertible to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr(qml_owned_ptr<U>&& u) noexcept
+            : m_ptr{u.m_ptr} {
+        u.m_ptr = nullptr;
+    }
+
+    // If U* is convertible to T* then qml_owned_ptr<U> is assignable to qml_owned_ptr<T>
+    template<
+            typename U,
+            typename = typename std::enable_if<std::is_convertible<U*, T*>::value, U>::type>
+    qml_owned_ptr& operator=(qml_owned_ptr<U>&& u) noexcept {
+        qml_owned_ptr temp{std::move(u)};
+        std::swap(temp.m_ptr, m_ptr);
+        DEBUG_ASSERT(!m_ptr ||
+                QQmlEngine::objectOwnership(m_ptr) ==
+                        QQmlEngine::JavaScriptOwnership);
+        return *this;
+    }
+
+    qml_owned_ptr& operator=(std::nullptr_t) noexcept {
+        qml_owned_ptr{std::move(*this)}; // move *this into a temporary that gets destructed
+        return *this;
```

Comment:
Isn't this a use-after-free? Did you mean to return `nullptr`?

Reply:
Same, it comes from [here](https://github.com/mixxxdj/mixxx/pull/2172/files#diff-3cca6b770fec14ac2dbf8d76d5e6b1549a079b1055fc4fa17718d069f1695907R60) - my understanding of this code is that `std:move` will move the wrapped ptr outside, but we still want to return a reference to the container that was referenced. In other term, `std:move` moves `m_ptr` outside of t`this` to `qml_owned_ptr.m_ptr`, but `this` remains allocated and valid.

But here again, I could be misunderstanding this code :)

Reply:
Ah yeah, no its fine.